webgl: max vectors limits

anholt opened the pull request:
This pull request fixes up some resource limits handling in angle, and requires an update of angle in the process to avoid regressions due to long expressions generated by the resource limits testcases.

This may or may not help #16220

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [x] These changes fix #20548
- [x] These changes fix #20549 

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19129)
<!-- Reviewable:end -->

anholt committed:
webgl: Set the resource limits for the shader validator.

This will get it to have the right constant values in shaders.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webglshader.rs
 * @fitzgen: components/script/dom/webglshader.rs
 * @emilio: components/canvas/webgl_thread.rs, components/script/dom/webglshader.rs

emilio commented:
r? @MortimerGoro 

You should be doing some reviews more often ;)

emilio commented:
@bors-servo delegate=MortimerGoro

bors-servo commented:
:v: @MortimerGoro can now approve this pull request

MortimerGoro commented:
Hi @anholt, thanks for doing this!

The problem I see is that it adds overhead when compiling shaders (a bunch of sync channel calls). Let's cache the shader validation params once at context setup as you say in one of the comments. There is already a place to do that:

Check limits property in `WebGLCreateContextResult`in
https://github.com/servo/servo/blob/master/components/canvas_traits/webgl.rs#L59

This GLLimits struct is defined in:
https://github.com/emilio/rust-offscreen-rendering-context/blob/master/src/gl_limits.rs

If we want to keep using that struct it will require a PR to rust-offscreen-rendering-context. Then you call pass a reference of the struct to shader.compile() and correctly set angle parameters.

anholt committed:
webgl: Query limits instead of pulling them from offscreen_gl_context.

offscreen_gl_context really doesn't have any reason to be caching all
of the limits that we're going to need in WebGL, so drop our use of
their cache and start doing it ourselves at context creation time.

anholt commented:
OK, the refactor wasn't as bad as I was thinking it might be.  Pushed a new version I'm pretty happy with.

MortimerGoro commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -177,6 +177,44 @@ impl TextureUnitBindings {
+        };
+
+        WebGLRenderingContextLimits {
+            max_vertex_attribs: get_param(constants::MAX_VERTEX_ATTRIBS),
```
Could you initialize this struct in the WebGLThread when a context is created?. I'd like to also avoid the getter channel steps when creating the context:  you have direct access to the context.gl() there in webgl_thread.rs. And send the result where the previous `limits` was used

anholt replied:
I'm trying, but it's such a layering violation that I'm having a hard time.

MortimerGoro replied:
You can define the struct in webgl.rs (canvas_traits component). That way it can be used in both layers without problems.

anholt replied:
The problem was then trying to get the constants back from the script component.  gleam::gl to the rescue.

MortimerGoro commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -231,12 +269,12 @@ impl WebGLRenderingContext {
 
         result.map(|ctx_data| {
             WebGLRenderingContext {
+                limits: WebGLRenderingContextLimits::new(&ctx_data.sender),
```
Get the WebGLRenderingContextLimits from ctx_data as mentioned in the other comment

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19192) made this pull request unmergeable. Please resolve the merge conflicts.

anholt commented:
Branch updated.  Not actually tested, because after the rebase my release build is broken with:

```
error: failed to run custom build command for `gecko-media v0.1.0 (https://github.com/servo/gecko-media.git#0299b455)`
process didn't exit successfully: `/home/anholt/src/servo/servo/target/release/build/gecko-media-59486aef098dd7b4/build-script-build` (exit code: 101)

[...]

--- stderr
gecko/glue/include/GeckoMedia.h:10:10: fatal error: 'stddef.h' file not found
gecko/glue/include/GeckoMedia.h:10:10: fatal error: 'stddef.h' file not found, err: true
thread 'main' panicked at 'Unable to generate bindings: ()', /checkout/src/libcore/result.rs:906:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

jdm commented:
I've filed https://github.com/servo/gecko-media/issues/71 about that.

MortimerGoro commented on components/canvas_traits/webgl.rs:
```diff
@@ -546,3 +547,36 @@ impl fmt::Debug for WebGLCommand {
+}
+
+impl WebGLContextLimits {
+    pub fn new(gl: &gl::Gl) -> WebGLContextLimits {
```
Thanks, great work!. Just one more thing, could you move the initialization of the struct to webgl_thread.rs (e.g.: to a separate factory method from the struct) ? I'd like to avoid adding gleam dependency to canvas_traits and doing gl code there. We are close from even removing offscreen_gl_context dependency from canvas_traits (on a separate PR I  mean). That will allow much faster compilations (e.g. from almost 10 minutos to less than 1) when developing & testing local changes in gleam & offscreen_gl because the traits are not recompiled which causes script component recompilation.

I'll merge the PR after doing that ;)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19270) made this pull request unmergeable. Please resolve the merge conflicts.

anholt commented:
Updated with the move of the gleam calls

MortimerGoro commented:
Thanks for doing this!

@bors-servo r+

bors-servo commented:
:pushpin: Commit e4c928a has been approved by `MortimerGoro`

<!-- @bors-servo r=MortimerGoro e4c928a3b0089c745480d9dc29767741917ab717 -->

bors-servo commented:
:hourglass: Testing commit e4c928a3b0089c745480d9dc29767741917ab717 with merge ca28b42176a94d07581eb0cd2faf9b0a4006b3ee...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/2786)

MortimerGoro commented:
```
--- stderr
error: Connection to server timed out
```

@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/9931) are reusable. Rebuilding only [arm32](http://build.servo.org/builders/arm32/builds/10167), [arm64](http://build.servo.org/builders/arm64/builds/10196), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10216), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7046), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7073), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10275), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/2708), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2723), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/6968), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7032), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/2786), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/2745), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5590)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7046)

MortimerGoro commented:
```
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "Overall test", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/glsl/bugs/complex-glsl-does-not-crash.html", "line": 146736, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/webgl/conformance-1.0.3/conformance/glsl/bugs/complex-glsl-does-not-crash.html", "line": 146779, "action": "test_result", "expected": "OK"}
```

anholt commented:
I suspect that might be an intermittent?  It's not reproducing here, and I don't think I should have changed its behavior.

MortimerGoro commented:
Yep, it looks like an intermittent. Let's try again

@bors-servo retry

bors-servo commented:
:hourglass: Testing commit e4c928a3b0089c745480d9dc29767741917ab717 with merge c1757922502ceda8d2c99c461ed265254f8b2410...

anholt commented:
I ran two more rounds of tests here, and I've seen the first and second lines of:

```
/_mozilla/webgl/conformance-1.0.3/conformance/glsl/misc/shader-uniform-packing-restrictions.html
------------------------------------------------------------------------------------------------
FAIL WebGL test #116: [unexpected link status] fragment shader with uniform array of ivec2 with 4097 elements (one past maximum) accessing last element should fail
FAIL WebGL test #117: [unexpected link status] fragment shader with uniform array of ivec2 with 4097 elements (one past maximum) accessing first element should fail
```

So we may have some more fails to re-mark in that test.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7035)

MortimerGoro commented:
Same error again

jdm commented:
When the test times out and there's NOTRUN for "Overall test", that usually indicates that there are just too many test cases for Servo to reliably finish running before the test harness timeout is reached.

anholt commented:
Never mind, this series definitely affects that testcase, and it is super slow to run now.  (looks like javascript overhead, unless I'm misusing perf)

anholt commented:
can we get a retest on this?

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 613430fc41d237102751c6c65d5824d4001b7c3f with merge a7bce64ffb68dc25df8bc883b06abeed084512d6...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7227)

anholt commented:
So now we're failing due to complex-glsl-does-not-crash *not* timing out?  Did I need some different syntax to allow success instead of timeout there to not count as an unexpected change?

jdm commented:
We don't have any support for tests that sometimes time out and sometimes pass, unfortunately. If the result is intermittent, we just end up filing an intermittent issue to track it and set the expected result as the most common one.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19728) made this pull request unmergeable. Please resolve the merge conflicts.

anholt committed:
update all the wpts

anholt committed:
webgl: Add getParameter() support for the limits we're caching.

anholt commented:
Massive rebase today, and a new change to add a pile of `getParameter()`s so that `glsl-built-ins.html.ini` can go away.

jdm commented:
Sigh. This fell on the floor and I independently reimplemented it in #21026. Given the rebasing that would be required to make this fit in, I'm going to close it in favour of #21026. Sorry @anholt :(
